### ISSUE TITLE
dependency failed to start

### ISSUE THREAD

**Issue by conradwt** [prometheus/prometheus#12910]
### What did you do?

1. docker compose up

### What did you expect to see?

I would expect the container to start without error.

### What did you see instead? Under which circumstances?

dependency failed to start: container prometheus is unhealthy

### System information

macOS 14 (Sonoma)

### Prometheus version

```text
v2.47.0
```

### Prometheus configuration file

```yaml
# my global config
global:
  scrape_interval: 15s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 15s # Evaluate rules every 15 seconds. The default is every 1 minute.
  # scrape_timeout is set to the global default (10s).

  external_labels:
    monitor: 'phoenix_app'

# Alertmanager configuration
alerting:
  alertmanagers:
    - static_configs:
        - targets:
          # - alertmanager:9093

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"

# A example scrape configuration for running Prometheus with Docker.

scrape_configs:
  # Make Prometheus scrape itself for metrics.
  - job_name: 'prometheus'
    static_configs:
      - targets: ['host.docker.internal:9090']
      # - targets: ['prometheus:9090']

  # Create a job for Docker daemon.
  #
  # This example requires Docker daemon to be configured to expose
  # Prometheus metrics, as documented here:
  # https://docs.docker.com/config/daemon/prometheus/
  - job_name: 'docker'
    static_configs:
      # - targets: ['127.0.0.1:9323']
      - targets: ['host.docker.internal:9323']

  #
  # Create a job for the Phoenix application.
  #
  # Reference:
  #
  # https://github.com/akoutmos/prom_ex/blob/master/example_applications/shared_docker/prometheus/config.yml#L15-L17
  #
  - job_name: 'phoenix_app'
    scrape_interval: 5s
    static_configs:
      - targets: ['host.docker.internal:4000']

  # Create a job for Docker Swarm containers.
  #
  # This example works with cadvisor running using:
  # docker run --detach --name cadvisor -l prometheus-job=cadvisor
  #     --mount type=bind,src=/var/run/docker.sock,dst=/var/run/docker.sock,ro
  #     --mount type=bind,src=/,dst=/rootfs,ro
  #     --mount type=bind,src=/var/run,dst=/var/run
  #     --mount type=bind,src=/sys,dst=/sys,ro
  #     --mount type=bind,src=/var/lib/docker,dst=/var/lib/docker,ro
  #     google/cadvisor -docker_only
  # - job_name: 'docker-containers'
  #   docker_sd_configs:
  #     - host: unix:///var/run/docker.sock # You can also use http/https to connect to the Docker daemon.
  #   relabel_configs:
  #     # Only keep containers that have a `prometheus-job` label.
  #     - source_labels: [__meta_docker_container_label_prometheus_job]
  #       regex: .+
  #       action: keep
  #     # Use the task labels that are prefixed by `prometheus-`.
  #     - regex: __meta_docker_container_label_prometheus_(.+)
  #       action: labelmap
  #       replacement: $1
# remote_write:
#   - url: '<Your Prometheus remote_write endpoint>'
#     basic_auth:
#       username: '<Your Grafana Username>'
#       password: '<Your Grafana API key>'
```

### Alertmanager version

```text
N/A
```

### Alertmanager configuration file

```yaml
N/A
```

### Logs

```text
docker compose up
[+] Running 2/0
 ✔ Container prometheus  Created                                                                                                                                                                              0.0s 
 ✔ Container grafana     Created                                                                                                                                                                              0.0s 
Attaching to grafana, prometheus
prometheus  | ts=2023-10-01T06:08:38.909Z caller=main.go:583 level=info msg="Starting Prometheus Server" mode=server version="(version=2.47.0, branch=HEAD, revision=efa34a5840661c29c2e362efa76bc3a70dccb335)"
prometheus  | ts=2023-10-01T06:08:38.909Z caller=main.go:588 level=info build_context="(go=go1.21.0, platform=linux/arm64, user=root@409eb5e6b30c, date=20230906-10:27:30, tags=netgo,builtinassets,stringlabels)"
prometheus  | ts=2023-10-01T06:08:38.909Z caller=main.go:589 level=info host_details="(Linux 6.3.13-linuxkit #1 SMP PREEMPT Thu Sep  7 07:48:47 UTC 2023 aarch64 39d6d04f566a (none))"
prometheus  | ts=2023-10-01T06:08:38.909Z caller=main.go:590 level=info fd_limits="(soft=1048576, hard=1048576)"
prometheus  | ts=2023-10-01T06:08:38.909Z caller=main.go:591 level=info vm_limits="(soft=unlimited, hard=unlimited)"
prometheus  | ts=2023-10-01T06:08:38.910Z caller=web.go:566 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
prometheus  | ts=2023-10-01T06:08:38.910Z caller=main.go:1024 level=info msg="Starting TSDB ..."
prometheus  | ts=2023-10-01T06:08:38.911Z caller=tls_config.go:274 level=info component=web msg="Listening on" address=[::]:9090
prometheus  | ts=2023-10-01T06:08:38.911Z caller=tls_config.go:277 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
prometheus  | ts=2023-10-01T06:08:38.912Z caller=head.go:600 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
prometheus  | ts=2023-10-01T06:08:38.912Z caller=head.go:681 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=1.042µs
prometheus  | ts=2023-10-01T06:08:38.912Z caller=head.go:689 level=info component=tsdb msg="Replaying WAL, this may take a while"
prometheus  | ts=2023-10-01T06:08:38.914Z caller=head.go:760 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=1
prometheus  | ts=2023-10-01T06:08:38.914Z caller=head.go:760 level=info component=tsdb msg="WAL segment loaded" segment=1 maxSegment=1
prometheus  | ts=2023-10-01T06:08:38.914Z caller=head.go:797 level=info component=tsdb msg="WAL replay completed" checkpoint_replay_duration=23.375µs wal_replay_duration=1.628625ms wbl_replay_duration=41ns total_replay_duration=1.663791ms
prometheus  | ts=2023-10-01T06:08:38.915Z caller=main.go:1045 level=info fs_type=EXT4_SUPER_MAGIC
prometheus  | ts=2023-10-01T06:08:38.915Z caller=main.go:1048 level=info msg="TSDB started"
prometheus  | ts=2023-10-01T06:08:38.915Z caller=main.go:1229 level=info msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml
prometheus  | ts=2023-10-01T06:08:38.916Z caller=main.go:1266 level=info msg="Completed loading of configuration file" filename=/etc/prometheus/prometheus.yml totalDuration=708.334µs db_storage=959ns remote_storage=1µs web_handler=334ns query_engine=5.417µs scrape=161.041µs scrape_sd=19.542µs notify=9.792µs notify_sd=3.125µs rules=834ns tracing=1.75µs
prometheus  | ts=2023-10-01T06:08:38.916Z caller=main.go:1009 level=info msg="Server is ready to receive web requests."
prometheus  | ts=2023-10-01T06:08:38.916Z caller=manager.go:1009 level=info component="rule manager" msg="Starting rule manager..."
dependency failed to start: container prometheus is unhealthy
```

**Comment by bwplotka** [comment]
Hi!

Could you give us more info on your compose configuration? We don't provide anything in this repo, so it's unclear how you run & configure Prometheus (other than scrape config). You injected some commentary about `# Create a job for Docker Swarm containers.` but it's within scrape config, so it's hard to parse, plus we don't see how you spin up Prometheus containers. Thanks! 🤗 

It feels you might missed some readiness probe (or healthcheck it's called in Docker swarm?)

(Triaged during Prometheus bug scrub meeting)

**Comment by conradwt** [comment]
@bwplotka I have created a sample repository that you can find at the following location:

https://github.com/conradwt/promex-example-using-docker-desktop

**Comment by machine424** [comment]
Yep, it looks like a docker compose error https://github.com/docker/compose/blob/2945532f978ba1fc5b4e448359cf18f2e35f76eb/pkg/compose/convergence.go#L375C1-L375C63

I suggest you take a look at those healthchecks logs (maybe `docker inspect` may help)

**Comment by beorn7** [comment]
Hello from the bug scrub!

We hope that the above pointers helped with the issue. After 1½ years of inactivity, we will close this.
